Multiple "index" files in a entrypoint directory

### Describe the bug

When you have two `index.*` files inside an entrypoint directory, like the Solid template, WXT will fail to build the application.

```
entrypoints/
  popup/
    index.html
    index.ts
    ...
```

WXT should not allow more than 1 entrypoint in a directory.

### Expected behavior

In this case, I'd either:

1. Only expect the HTML file to be counted as an entrypoint, and the `index.ts` file would be ignored.
2. Expect to be told I can't have two `index` files in a directory
   - Perhaps we need to move the "no duplicate entrypoint name" validation up before we import the files to grab their config.

### Screenshots

Here is one such failure:

```
[log] WXT 0.11.1
The CJS build of Vite's Node API is deprecated. See https://vitejs.dev/guide/troubleshooting.html#vite-cjs-node-api-deprecated for more details.
[info] Building chrome-mv3 for production with Vite 5.0.8
Error:  Cannot find module 'react/jsx-runtime'
Require stack:
- /home/runner/work/wxt/wxt/templates/solid/entrypoints/popup/index.tsx
  Require stack:
  - entrypoints/popup/index.tsx
  at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
  at Function.resolve (node:internal/modules/helpers:188:19)
  at _resolve (node_modules/jiti/dist/jiti.js:1:251148)
  at jiti (node_modules/jiti/dist/jiti.js:1:253746)
  at entrypoints/popup/index.tsx:1:88
  at evalModule (node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (node_modules/jiti/dist/jiti.js:1:254371)
  at importEntrypointFile (node_modules/wxt/dist/cli.cjs:4163:23)
  at async getUnlistedScriptEntrypoint (node_modules/wxt/dist/cli.cjs:2798:25)
  at async node_modules/wxt/dist/cli.cjs:2648:18
Error:  Cannot find module 'react/jsx-runtime'
[fail] Command failed after 642 ms
Require stack:
- /home/runner/work/wxt/wxt/templates/solid/entrypoints/popup/index.tsx
  Require stack:
  - entrypoints/popup/index.tsx
  at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
  at Function.resolve (node:internal/modules/helpers:188:19)
  at _resolve (node_modules/jiti/dist/jiti.js:1:251148)
  at jiti (node_modules/jiti/dist/jiti.js:1:253746)
  at entrypoints/popup/index.tsx:1:88
  at evalModule (node_modules/jiti/dist/jiti.js:1:256443)
  at jiti (node_modules/jiti/dist/jiti.js:1:254371)
  at importEntrypointFile (node_modules/wxt/dist/cli.cjs:4163:23)
  at async getUnlistedScriptEntrypoint (node_modules/wxt/dist/cli.cjs:2798:25)
  at async node_modules/wxt/dist/cli.cjs:2648:18
 ELIFECYCLE  Command failed with exit code 1.
Error: Process completed with exit code 1.
```

### Environment

<!--- Run `npx envinfo --system --browsers --binaries --npmPackages wxt,vite` and paste the output below -->

```
  System:
    OS: macOS 13.4.1
    CPU: (8) arm64 Apple M1 Pro
    Memory: 3.11 GB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 18.16.1 - ~/.asdf/installs/nodejs/18.16.1/bin/node
    npm: 9.5.1 - ~/.asdf/plugins/nodejs/shims/npm
    pnpm: 8.6.3 - ~/.asdf/installs/nodejs/18.16.1/bin/pnpm
  Browsers:
    Chrome: 120.0.6099.71
    Safari: 16.5.2
  npmPackages:
    wxt: ^0.11.0 => 0.11.1 
```

### Additional context

Discovered while working on #274.

--- a/src/core/utils/building/__tests__/find-entrypoints.test.ts
+++ b/src/core/utils/building/__tests__/find-entrypoints.test.ts
@@ -569,8 +569,13 @@ describe('findEntrypoints', () => {
     ].map(unnormalizePath);
 
     await expect(() => findEntrypoints(config)).rejects.toThrowError(
-      'Multiple entrypoints with the name "popup" detected, but only one is allowed: ' +
-        expectedPaths.join(', '),
+      [
+        'Multiple entrypoints with the same name detected, only one entrypoint for each name is allowed.',
+        '',
+        '- popup',
+        `  - ${unnormalizePath('src/entrypoints/popup.html')}`,
+        `  - ${unnormalizePath('src/entrypoints/popup/index.html')}`,
+      ].join('\n'),
     );
   });
 

--- a/src/core/utils/building/__tests__/find-entrypoints.test.ts
+++ b/src/core/utils/building/__tests__/find-entrypoints.test.ts
@@ -562,7 +562,11 @@ describe('findEntrypoints', () => {
   });
 
   it('should not allow multiple entrypoints with the same name', async () => {
-    globMock.mockResolvedValueOnce(['popup.html', 'popup/index.html']);
+    globMock.mockResolvedValueOnce([
+      'popup.html',
+      'popup/index.html',
+      'popup/index.ts',
+    ]);
 
     await expect(() => findEntrypoints(config)).rejects.toThrowError(
       [
@@ -571,6 +575,7 @@ describe('findEntrypoints', () => {
         '- popup',
         `  - ${unnormalizePath('src/entrypoints/popup.html')}`,
         `  - ${unnormalizePath('src/entrypoints/popup/index.html')}`,
+        `  - ${unnormalizePath('src/entrypoints/popup/index.ts')}`,
       ].join('\n'),
     );
   });

--- a/src/core/utils/building/__tests__/find-entrypoints.test.ts
+++ b/src/core/utils/building/__tests__/find-entrypoints.test.ts
@@ -563,15 +563,21 @@ describe('findEntrypoints', () => {
 
   it('should not allow multiple entrypoints with the same name', async () => {
     globMock.mockResolvedValueOnce([
+      'options/index.html',
+      'options/index.jsx',
       'popup.html',
       'popup/index.html',
       'popup/index.ts',
+      'ui.html',
     ]);
 
     await expect(() => findEntrypoints(config)).rejects.toThrowError(
       [
         'Multiple entrypoints with the same name detected, only one entrypoint for each name is allowed.',
         '',
+        '- options',
+        `  - ${unnormalizePath('src/entrypoints/options/index.html')}`,
+        `  - ${unnormalizePath('src/entrypoints/options/index.jsx')}`,
         '- popup',
         `  - ${unnormalizePath('src/entrypoints/popup.html')}`,
         `  - ${unnormalizePath('src/entrypoints/popup/index.html')}`,

--- a/src/core/utils/building/__tests__/find-entrypoints.test.ts
+++ b/src/core/utils/building/__tests__/find-entrypoints.test.ts
@@ -563,10 +563,6 @@ describe('findEntrypoints', () => {
 
   it('should not allow multiple entrypoints with the same name', async () => {
     globMock.mockResolvedValueOnce(['popup.html', 'popup/index.html']);
-    const expectedPaths = [
-      'src/entrypoints/popup.html',
-      'src/entrypoints/popup/index.html',
-    ].map(unnormalizePath);
 
     await expect(() => findEntrypoints(config)).rejects.toThrowError(
       [

--- a/src/core/utils/building/find-entrypoints.ts
+++ b/src/core/utils/building/find-entrypoints.ts
@@ -1,4 +1,4 @@
-import { resolve, relative } from 'path';
+import { relative, resolve } from 'path';
 import {
   BackgroundEntrypoint,
   BackgroundDefinition,
@@ -46,15 +46,7 @@ export async function findEntrypoints(
     const matchingGlob = pathGlobs.find((glob) =>
       minimatch(relativePath, glob),
     );
-    if (matchingGlob == null) {
-      config.logger.warn(
-        `${relativePath} does not match any known entrypoint. Known entrypoints:\n${JSON.stringify(
-          PATH_GLOB_TO_TYPE_MAP,
-          null,
-          2,
-        )}`,
-      );
-    } else {
+    if (matchingGlob) {
       const type = PATH_GLOB_TO_TYPE_MAP[matchingGlob];
       results.push({ name, inputPath, type });
     }
@@ -62,7 +54,7 @@ export async function findEntrypoints(
   }, []);
 
   // Report duplicate entrypoint names
-  validateEntrypointNames(config, entrypointInfos);
+  preventDuplicateEntrypointNames(config, entrypointInfos);
 
   // Import entrypoints to get their config
   let hasBackground = false;
@@ -145,11 +137,10 @@ interface EntrypointInfo {
   type: Entrypoint['type'];
 }
 
-function validateEntrypointNames(
+function preventDuplicateEntrypointNames(
   config: InternalConfig,
   files: EntrypointInfo[],
 ) {
-  // No duplicate names
   const namesToPaths = files.reduce<Record<string, string[]>>(
     (map, { name, inputPath }) => {
       map[name] ??= [];

--- a/src/core/utils/building/find-entrypoints.ts
+++ b/src/core/utils/building/find-entrypoints.ts
@@ -1,4 +1,4 @@
-import { relative, resolve } from 'path';
+import { resolve, relative } from 'path';
 import {
   BackgroundEntrypoint,
   BackgroundDefinition,
@@ -31,56 +31,62 @@ import { CSS_EXTENSIONS_PATTERN } from '~/core/utils/paths';
 export async function findEntrypoints(
   config: InternalConfig,
 ): Promise<Entrypoint[]> {
-  const relativePaths = await glob('**/*', {
+  const relativePaths = await glob(Object.keys(PATH_GLOB_TO_TYPE_MAP), {
     cwd: config.entrypointsDir,
   });
   // Ensure consistent output
   relativePaths.sort();
 
   const pathGlobs = Object.keys(PATH_GLOB_TO_TYPE_MAP);
-
-  let hasBackground = false;
-  // TODO: This parallelization is bad
-  const possibleEntrypoints: Array<Entrypoint | undefined> = await Promise.all(
-    relativePaths.map(async (relativePath) => {
-      const path = resolve(config.entrypointsDir, relativePath);
-      const matchingGlob = pathGlobs.find((glob) =>
-        minimatch(relativePath, glob),
+  const entrypointInfos: EntrypointInfo[] = relativePaths.reduce<
+    EntrypointInfo[]
+  >((results, relativePath) => {
+    const inputPath = resolve(config.entrypointsDir, relativePath);
+    const name = getEntrypointName(config.entrypointsDir, inputPath);
+    const matchingGlob = pathGlobs.find((glob) =>
+      minimatch(relativePath, glob),
+    );
+    if (matchingGlob == null) {
+      config.logger.warn(
+        `${relativePath} does not match any known entrypoint. Known entrypoints:\n${JSON.stringify(
+          PATH_GLOB_TO_TYPE_MAP,
+          null,
+          2,
+        )}`,
       );
-
-      if (matchingGlob == null) {
-        config.logger.warn(
-          `${relativePath} does not match any known entrypoint. Known entrypoints:\n${JSON.stringify(
-            PATH_GLOB_TO_TYPE_MAP,
-            null,
-            2,
-          )}`,
-        );
-        return;
-      }
-
+    } else {
       const type = PATH_GLOB_TO_TYPE_MAP[matchingGlob];
-      if (type === 'ignored') return;
+      results.push({ name, inputPath, type });
+    }
+    return results;
+  }, []);
 
+  // Report duplicate entrypoint names
+  validateEntrypointNames(config, entrypointInfos);
+
+  // Import entrypoints to get their config
+  let hasBackground = false;
+  const entrypoints: Entrypoint[] = await Promise.all(
+    entrypointInfos.map(async (info): Promise<Entrypoint> => {
+      const { type } = info;
       switch (type) {
         case 'popup':
-          return await getPopupEntrypoint(config, path);
+          return await getPopupEntrypoint(config, info);
         case 'options':
-          return await getOptionsEntrypoint(config, path);
+          return await getOptionsEntrypoint(config, info);
         case 'background':
           hasBackground = true;
-          return await getBackgroundEntrypoint(config, path);
+          return await getBackgroundEntrypoint(config, info);
         case 'content-script':
-          return await getContentScriptEntrypoint(config, path);
+          return await getContentScriptEntrypoint(config, info);
         case 'unlisted-page':
-          return await getUnlistedPageEntrypoint(config, path);
+          return await getUnlistedPageEntrypoint(config, info);
         case 'unlisted-script':
-          return await getUnlistedScriptEntrypoint(config, path);
+          return await getUnlistedScriptEntrypoint(config, info);
         case 'content-script-style':
           return {
+            ...info,
             type,
-            name: getEntrypointName(config.entrypointsDir, path),
-            inputPath: path,
             outputDir: resolve(config.outDir, CONTENT_SCRIPT_OUT_DIR),
             options: {
               include: undefined,
@@ -89,9 +95,8 @@ export async function findEntrypoints(
           };
         default:
           return {
+            ...info,
             type,
-            name: getEntrypointName(config.entrypointsDir, path),
-            inputPath: path,
             outputDir: config.outDir,
             options: {
               include: undefined,
@@ -102,30 +107,13 @@ export async function findEntrypoints(
     }),
   );
 
-  const entrypoints = possibleEntrypoints.filter(
-    (entry) => !!entry,
-  ) as Entrypoint[];
-
-  // Report duplicate entrypoint names
-  const existingNames: Record<string, Entrypoint | undefined> = {};
-  entrypoints.forEach((entrypoint) => {
-    const withSameName = existingNames[entrypoint.name];
-    if (withSameName) {
-      throw Error(
-        `Multiple entrypoints with the name "${
-          entrypoint.name
-        }" detected, but only one is allowed: ${[
-          relative(config.root, withSameName.inputPath),
-          relative(config.root, entrypoint.inputPath),
-        ].join(', ')}`,
-      );
-    }
-    existingNames[entrypoint.name] = entrypoint;
-  });
-
   if (config.command === 'serve' && !hasBackground) {
     entrypoints.push(
-      await getBackgroundEntrypoint(config, VIRTUAL_NOOP_BACKGROUND_MODULE_ID),
+      await getBackgroundEntrypoint(config, {
+        inputPath: VIRTUAL_NOOP_BACKGROUND_MODULE_ID,
+        name: 'background',
+        type: 'background',
+      }),
     );
   }
 
@@ -149,6 +137,45 @@ export async function findEntrypoints(
   });
   config.logger.debug(`${config.browser} entrypoints:`, targetEntrypoints);
   return targetEntrypoints;
+}
+
+interface EntrypointInfo {
+  name: string;
+  inputPath: string;
+  type: Entrypoint['type'];
+}
+
+function validateEntrypointNames(
+  config: InternalConfig,
+  files: EntrypointInfo[],
+) {
+  // No duplicate names
+  const namesToPaths = files.reduce<Record<string, string[]>>(
+    (map, { name, inputPath }) => {
+      map[name] ??= [];
+      map[name].push(inputPath);
+      return map;
+    },
+    {},
+  );
+  const errorLines = Object.entries(namesToPaths).reduce<string[]>(
+    (lines, [name, absolutePaths]) => {
+      if (absolutePaths.length > 1) {
+        lines.push(`- ${name}`);
+        absolutePaths.forEach((absolutePath) => {
+          lines.push(`  - ${relative(config.root, absolutePath)}`);
+        });
+      }
+      return lines;
+    },
+    [],
+  );
+  if (errorLines.length > 0) {
+    const errorContent = errorLines.join('\n');
+    throw Error(
+      `Multiple entrypoints with the same name detected, only one entrypoint for each name is allowed.\n\n${errorContent}`,
+    );
+  }
 }
 
 function getHtmlBaseOptions(document: Document): BaseEntrypointOptions {
@@ -177,9 +204,9 @@ function getHtmlBaseOptions(document: Document): BaseEntrypointOptions {
  */
 async function getPopupEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<PopupEntrypoint> {
-  const content = await fs.readFile(path, 'utf-8');
+  const content = await fs.readFile(inputPath, 'utf-8');
   const { document } = parseHTML(content);
 
   const options: PopupEntrypoint['options'] = getHtmlBaseOptions(document);
@@ -220,7 +247,7 @@ async function getPopupEntrypoint(
     type: 'popup',
     name: 'popup',
     options,
-    inputPath: path,
+    inputPath,
     outputDir: config.outDir,
   };
 }
@@ -231,9 +258,9 @@ async function getPopupEntrypoint(
  */
 async function getOptionsEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<OptionsEntrypoint> {
-  const content = await fs.readFile(path, 'utf-8');
+  const content = await fs.readFile(inputPath, 'utf-8');
   const { document } = parseHTML(content);
 
   const options: OptionsEntrypoint['options'] = getHtmlBaseOptions(document);
@@ -263,7 +290,7 @@ async function getOptionsEntrypoint(
     type: 'options',
     name: 'options',
     options,
-    inputPath: path,
+    inputPath,
     outputDir: config.outDir,
   };
 }
@@ -274,15 +301,15 @@ async function getOptionsEntrypoint(
  */
 async function getUnlistedPageEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<GenericEntrypoint> {
-  const content = await fs.readFile(path, 'utf-8');
+  const content = await fs.readFile(inputPath, 'utf-8');
   const { document } = parseHTML(content);
 
   return {
     type: 'unlisted-page',
-    name: getEntrypointName(config.entrypointsDir, path),
-    inputPath: path,
+    name: getEntrypointName(config.entrypointsDir, inputPath),
+    inputPath,
     outputDir: config.outDir,
     options: getHtmlBaseOptions(document),
   };
@@ -294,11 +321,10 @@ async function getUnlistedPageEntrypoint(
  */
 async function getUnlistedScriptEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<GenericEntrypoint> {
-  const name = getEntrypointName(config.entrypointsDir, path);
   const defaultExport = await importEntrypointFile<UnlistedScriptDefinition>(
-    path,
+    inputPath,
     config,
   );
   if (defaultExport == null) {
@@ -311,7 +337,7 @@ async function getUnlistedScriptEntrypoint(
   return {
     type: 'unlisted-script',
     name,
-    inputPath: path,
+    inputPath,
     outputDir: config.outDir,
     options,
   };
@@ -322,13 +348,12 @@ async function getUnlistedScriptEntrypoint(
  */
 async function getBackgroundEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<BackgroundEntrypoint> {
-  const name = 'background';
   let options: Omit<BackgroundDefinition, 'main'> = {};
-  if (path !== VIRTUAL_NOOP_BACKGROUND_MODULE_ID) {
+  if (inputPath !== VIRTUAL_NOOP_BACKGROUND_MODULE_ID) {
     const defaultExport = await importEntrypointFile<BackgroundDefinition>(
-      path,
+      inputPath,
       config,
     );
     if (defaultExport == null) {
@@ -342,7 +367,7 @@ async function getBackgroundEntrypoint(
   return {
     type: 'background',
     name,
-    inputPath: path,
+    inputPath,
     outputDir: config.outDir,
     options: {
       ...options,
@@ -357,11 +382,10 @@ async function getBackgroundEntrypoint(
  */
 async function getContentScriptEntrypoint(
   config: InternalConfig,
-  path: string,
+  { inputPath, name }: EntrypointInfo,
 ): Promise<ContentScriptEntrypoint> {
-  const name = getEntrypointName(config.entrypointsDir, path);
   const { main: _, ...options } =
-    await importEntrypointFile<ContentScriptDefinition>(path, config);
+    await importEntrypointFile<ContentScriptDefinition>(inputPath, config);
   if (options == null) {
     throw Error(
       `${name}: Default export not found, did you forget to call "export default defineContentScript(...)"?`,
@@ -370,13 +394,13 @@ async function getContentScriptEntrypoint(
   return {
     type: 'content-script',
     name,
-    inputPath: path,
+    inputPath,
     outputDir: resolve(config.outDir, CONTENT_SCRIPT_OUT_DIR),
     options,
   };
 }
 
-const PATH_GLOB_TO_TYPE_MAP: Record<string, Entrypoint['type'] | 'ignored'> = {
+const PATH_GLOB_TO_TYPE_MAP: Record<string, Entrypoint['type']> = {
   'sandbox.html': 'sandbox',
   'sandbox/index.html': 'sandbox',
   '*.sandbox.html': 'sandbox',
@@ -424,9 +448,6 @@ const PATH_GLOB_TO_TYPE_MAP: Record<string, Entrypoint['type'] | 'ignored'> = {
   '*/index.[jt]s?(x)': 'unlisted-script',
   [`*.${CSS_EXTENSIONS_PATTERN}`]: 'unlisted-style',
   [`*/index.${CSS_EXTENSIONS_PATTERN}`]: 'unlisted-style',
-
-  // Don't warn about any files in subdirectories, like CSS or JS entrypoints for HTML files or tests
-  '*/**': 'ignored',
 };
 
 const CONTENT_SCRIPT_OUT_DIR = 'content-scripts';

